sdks/python/apache_beam/testing/analyzers/perf_regression_analysis.py

```diff
@@ -0,0 +1,372 @@
+  BigQuery when a GitHub issue is created on a performance
+  alert.
+  """
+  issue_timestamp: pd.Timestamp
+  change_point_timestamp: pd.Timestamp
+  test_name: str
+  metric_name: str
+  issue_number: int
+  issue_url: str
+  test_id: str
+  change_point: float
+
+
+class ChangePointAnalysis:
+  def __init__(
+      self,
+      data: Union[List[float], List[List[float]], np.ndarray],
+      metric_name: str,
+  ):
+    self.data = data
+    self.metric_name = metric_name
+
+  def edivisive_means(self,
+                      pvalue: float = 0.05,
+                      permutations: int = 100) -> List[int]:
+    """
+    Args:
+     pvalue: p value for the permutation test.
+     permutations: Number of permutations for the permutation test.
+     For more information, please look at
+     https://pypi.org/project/signal-processing-algorithms/
+
+    Performs edivisive means on the data and returns the indices of the
+    Change points.
+
+    Returns:
+     The indices of change points.
+    """
+    return e_divisive(self.data, pvalue, permutations)
+
+
+def is_change_point_in_valid_window(
+    num_runs_in_change_point_window: int, change_point_index: int) -> bool:
+  # If the change point is more than N runs behind the most recent run,
+  # Ignore the change point and don't raise an alert for it.
+  return num_runs_in_change_point_window >= change_point_index
+
+
+def find_existing_issue(
+    metric_name: str,
+    test_name: str,
+    change_point_timestamp: pd.Timestamp,
+    sibling_change_point_min_timestamp: pd.Timestamp,
+    sibling_change_point_max_timestamp: pd.Timestamp,
+) -> Optional[Tuple[bool, Optional[int]]]:
+  """
+  Finds the most recent GitHub issue created for change points for this
+  test+metric in sibling change point min and max timestamps window.
+  Returns a boolean and an issue ID whether the issue needs to be updated.
+  """
+  query_template = f"""
+  SELECT * FROM {_BQ_PROJECT_NAME}.{_BQ_DATASET}.{test_name}
+  WHERE {METRIC_NAME} = '{metric_name}'
+  ORDER BY {ISSUE_CREATION_TIMESTAMP_LABEL} DESC
+  LIMIT 1
+  """
+  try:
+    df = BigQueryMetricsFetcher().get_metrics(
+        query_template=query_template,
+        limit=_NUM_DATA_POINTS_TO_RUN_CHANGE_POINT_ANALYSIS)
+  except exceptions.NotFound:
+    # If no table found, that means this is first performance regression
+    # on the current test+metric.
+    return True, None
+  issue_number = df[ISSUE_NUMBER].tolist()[0]
+
+  if (sibling_change_point_min_timestamp <= change_point_timestamp <=
+      sibling_change_point_max_timestamp):
+    return False, None
+  return True, issue_number
+
+
+def read_test_config(config_file_path: str) -> Dict:
+  """
+  Reads the config file in which the data required to
+  run the change point analysis is specified.
+  """
+  with open(config_file_path, 'r') as stream:
+    config = yaml.safe_load(stream)
+  return config
+
+
+def run(config_file_path: str = None) -> None:
```

Comment:
try to use helpers to break up this method for readability, as it is getting long and does many different things.

for example, it might look like so:

```
def run():
   tests = parse_config(...)
   for test in tests: 
       change_point = find_most_recent_change_point(test)
       if not_reported(change_point):
          report_change_point_on_issue_tracker(change_point) 
```